### PR TITLE
Add fork tests as Python unit tests

### PR DESCRIPTION
**Pull request by ericgribkoff**
This includes the fork interop test cases as part of the Python unit test suite. The test pre-fork a local test interop server, and set the polling strategy to epoll1 on Linux and poll on Mac.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,955,026      Total (=)      1,955,026

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,843,757      Total (>)     10,843,747

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,955,026      Total (=)      1,955,026

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,843,755      Total (<)     10,843,757

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,955,026      Total (=)      1,955,026

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,843,752      Total (>)     10,843,747

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,955,026      Total (=)      1,955,026

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,843,760      Total (>)     10,843,752

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,955,026      Total (=)      1,955,026

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,843,755      Total (>)     10,843,749

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,955,026      Total (=)      1,955,026

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,843,752      Total (>)     10,843,750

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,955,026      Total (=)      1,955,026

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,843,754      Total (>)     10,843,751

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,955,026      Total (=)      1,955,026

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,843,757      Total (>)     10,843,748

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,955,026      Total (=)      1,955,026

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,843,755      Total (>)     10,843,752

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,955,026      Total (=)      1,955,026

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,843,755      Total (>)     10,843,749

 No significant differences in binary sizes


```

**Comment by ericgribkoff** [comment]
@gnossen PTAL. The changes to core in the first commit can be ignored; it is under separate review at https://github.com/grpc/grpc/pull/17973

File: src/python/grpcio_tests/commands.py [hunk 1]
```diff
@@ -95,7 +95,7 @@ def run(self):
 
         import tests
         loader = tests.Loader()
-        loader.loadTestsFromNames(['tests'])
+        loader.loadTestsFromNames(['tests.fork._fork_interop_test'])
```
Comment (gnossen) [inline src/python/grpcio_tests/commands.py]: Intentional?

Comment (ericgribkoff) [reply src/python/grpcio_tests/commands.py]: Oops, no. Fixed.

**Comment by ericgribkoff** [comment]
@gnossen This should be (finally...) ready for review. I filed https://github.com/grpc/grpc/issues/18075 to track the issues with 3.7, and for now these tests will only run on Python 2 - but this is far better than nothing, as it would have caught the fork-support breakage fixed-after-the-fact by https://github.com/grpc/grpc/pull/17973 and does catch a breakage in the in-progress https://github.com/grpc/grpc/pull/17308.

File: src/python/grpcio_tests/tests/fork/client.py [hunk 2]
```diff
@@ -63,12 +63,12 @@ def _test_case_from_arg(test_case_arg):
 
 def test_fork():
     logging.basicConfig(level=logging.INFO)
-    args = _args()
-    if args.test_case == "all":
+    args = vars(_args())
```
Comment (gnossen) [inline src/python/grpcio_tests/tests/fork/client.py]: What was the intent of this change?

Edit: I see. Unified interface for both CLI and programmatic invocation. :+1:

File: src/python/grpcio_tests/tests/fork/methods.py [hunk 3]
```diff
@@ -213,54 +218,59 @@ def _close_channel_before_fork(channel, args):
...
-            raise ValueError('Received connectivity updates on parent callback')
-        child_channel.unsubscribe(child_connectivity_callback)
-        child_channel.close()
+        with _channel(args) as child_channel:
+            child_stub = test_pb2_grpc.TestServiceStub(child_channel)
+            child_channel.subscribe(child_connectivity_callback)
+            _async_unary(child_stub)
+            if not child_channel_ready_event.wait(timeout=_RPC_TIMEOUT_S):
+                raise ValueError('Channel did not move to READY')
+            if len(parent_states) > 1:
```
Comment (gnossen) [inline src/python/grpcio_tests/tests/fork/methods.py]: Can we initialize `parent_states` earlier in the file than its first reference?

File: src/python/grpcio_tests/tests/fork/methods.py [hunk 4]
```diff
@@ -213,54 +218,59 @@ def _close_channel_before_fork(channel, args):
...
 
 def _connectivity_watch(channel, args):
 
     def child_target():
 
+        child_channel_ready_event = threading.Event()
+
         def child_connectivity_callback(state):
-            child_states.append(state)
+            if state is grpc.ChannelConnectivity.READY:
```
Comment (gnossen) [inline src/python/grpcio_tests/tests/fork/methods.py]: Nit: Seems like there's some parallelism in how state changes are gathered in the child and in the parent that could be factored out.

Edit: I'm noticing now that there's actually asymmetry here. `child_connectivity_callback` never records the new state. In fact, `child_states` is initialized but never used.

File: src/python/grpcio_tests/tests/fork/methods.py [hunk 5]
```diff
@@ -213,54 +218,59 @@ def _close_channel_before_fork(channel, args):
...
-    child_process.finish()
+    with _channel(args) as new_channel:
+        new_stub = test_pb2_grpc.TestServiceStub(new_channel)
+        child_process = _ChildProcess(child_target)
+        child_process.start()
+        _blocking_unary(new_stub)
+        child_process.finish()
 
 
 def _connectivity_watch(channel, args):
```
Comment (gnossen) [inline src/python/grpcio_tests/tests/fork/methods.py]: Nit: Might want to change name from `channel` to `parent_channel`.

File: src/python/grpcio_tests/tests/fork/methods.py [hunk 6]
```diff
@@ -130,7 +134,9 @@ def start(self):
         self._process.start()
 
     def finish(self):
-        self._process.join()
+        self._process.join(timeout=_CHILD_FINISH_TIMEOUT_S)
+        if self._process.is_alive():
+            raise ValueError('Child process did not terminate')
```
Comment (gnossen) [inline src/python/grpcio_tests/tests/fork/methods.py]: [`ValueError` is meant for invalid input.](https://docs.python.org/2/library/exceptions.html#exceptions.ValueError) I'd recommend subclassing [`RuntimeError`](https://docs.python.org/2/library/exceptions.html#exceptions.RuntimeError).

File: src/python/grpcio_tests/tests/fork/_fork_interop_test.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+            server.start()
+            print(port)
+            sys.stdout.flush()
+            while True:
+                time.sleep(1)
+        """
+        self._server_process = subprocess.Popen(
+            [sys.executable, '-c', start_server_script],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE)
```
Comment (gnossen) [inline src/python/grpcio_tests/tests/fork/_fork_interop_test.py]: Return code check?

File: src/python/grpcio_tests/tests/fork/_fork_interop_test.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+# compatibility with multiprocessing, we first fork+exec a new process to ensure
+# we don't have any conflicting background threads.
+_CLIENT_FORK_SCRIPT_TEMPLATE = """if True:
+    import os
+    import sys
+    from grpc._cython import cygrpc
+    from tests.fork import methods
+
+    cygrpc._GRPC_ENABLE_FORK_SUPPORT = True
+    os.environ['GRPC_POLL_STRATEGY'] = 'epoll1'
```
Comment (gnossen) [inline src/python/grpcio_tests/tests/fork/_fork_interop_test.py]: What's the story on this setting?

File: src/python/grpcio/grpc/_cython/_cygrpc/fork_posix.pyx.pxi [hunk 9]
```diff
@@ -37,15 +36,20 @@ _GRPC_ENABLE_FORK_SUPPORT = (
     os.environ.get('GRPC_ENABLE_FORK_SUPPORT', '0')
         .lower() in _TRUE_VALUES)
 
+_fork_handler_failed = False
+
 cdef void __prefork() nogil:
     with gil:
+        global _fork_handler_failed
```
Comment (gnossen) [inline src/python/grpcio/grpc/_cython/_cygrpc/fork_posix.pyx.pxi]: Is it possible for two threads to call `__prefork` at once?

**Pull request merged**